## Part 1

In [1]:
data = """
3-5
10-14
16-20
12-18

1
5
8
11
17
32
""".strip()

with open("input.txt", "r") as f:
    data = f.read()

def get_parts(data:str):
    range_data, id_data = data.strip().split("\n\n")
    ranges = []

    for r in range_data.split("\n"):
        ranges.append(tuple([int(n) for n in r.split("-")]))
    ids = tuple([int(idx) for idx in id_data.split("\n")])

    return ranges, ids

def count_fresh(ranges, ids):
    counter = 0
    for idx in ids:
        for r_min, r_max in ranges:
            if idx >= r_min and idx <= r_max:
                counter += 1
                break
    return counter

count_fresh(*get_parts(data))

770

Attempts:
- ⭐️ 770

## Part 2

In [2]:
import logging

data = """
3-5
10-14
16-20
12-18

1
5
8
11
17
32
""".strip()

with open("input.txt", "r") as f:
    data = f.read()


log = logging.getLogger()
# logging.basicConfig(level=logging.DEBUG) # enable debug output

def make_tuple(items):
    return tuple([tuple(el) for el in items])

def merge_ranges(ranges):
    new_ranges = []
    for r_min, r_max in sorted(ranges):
        if not len(new_ranges):
            new_ranges.append([r_min, r_max])
            log.debug(f"➕ [{r_min}-{r_max}]")
            continue

        for new_range_idx in range(0, len(new_ranges)):
            nr_min, nr_max = new_ranges[new_range_idx]

            # drop when range is of one element and equal to either start of end of another range
            if r_min == r_max and (r_min == nr_min or r_min == nr_max):
                log.debug(f"🟰 [{r_min}-{r_max}][{nr_min}-{nr_max}] -> 🗑️")
                break

            # drop when range is equal to the existing range
            if r_min == nr_min and r_max == nr_max:
                log.debug(f"🟰🟰 [{r_min}-{r_max}][{nr_min}-{nr_max}] -> 🗑️")
                break

            # extend to right when old range overlaps with new range
            if r_min <= nr_max and r_max > nr_max:
                new_ranges[new_range_idx] = [nr_min, r_max]
                log.debug(f"➡️ [{r_min}-{r_max}][{nr_min}-{nr_max}] -> [{nr_min}-{r_max}]")
                break

            # extend to left is not needed because of sorting

            # drop range if it falls within existing range
            if r_min >= nr_min and r_max <= nr_max:
                log.debug(f"⬇️ [{r_min}-{r_max}][{nr_min}-{nr_max}] -> 🗑️")
                break
        else:
            # append range as it is unseen
            new_ranges.append([r_min, r_max])
            log.debug(f"➕ [{r_min}-{r_max}]")

    final = make_tuple(new_ranges)

    # the reduce might span across multiple rounds
    # only return the final ranges list when no more changes are done after the whole procedure
    if final == make_tuple(ranges):
        return final

    # when changes were done, apply the reduce once more
    return merge_ranges(final)

def calc_fresh_ingredients(ranges) -> int:
    counter = 0
    for r_min, r_max in ranges:
        counter += r_max - r_min + 1
    return counter

ranges, ids = get_parts(data)

calc_fresh_ingredients(merge_ranges(ranges))

357674099117260

Attempts:
- 407750414415285 your answer is too high
- 391932400064531
- 391932400064527
- 391532621069584 your answer is too high
- 371129043970986
- ⭐️ 357674099117260